ReplicatedMapReadYourWritesTest.testReadYourWritesByGet

Failed on commit https://github.com/hazelcast/hazelcast/pull/18067/commits/735c5564554b0dbf1c8bedfc0fb73d551c7800fc of the PR https://github.com/hazelcast/hazelcast/pull/18067

Failed on INSERT_PLATFORM_INFORMATION_HERE: [Jenkins](http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4825/testReport/junit/com.hazelcast.replicatedmap/ReplicatedMapReadYourWritesTest/testReadYourWritesByGet/)

Stacktrace:
```
java.lang.NullPointerException
	at com.hazelcast.replicatedmap.ReplicatedMapReadYourWritesTest.assertReadYourWriteByGet(ReplicatedMapReadYourWritesTest.java:104)
	at com.hazelcast.replicatedmap.ReplicatedMapReadYourWritesTest.testReadYourWritesByGet(ReplicatedMapReadYourWritesTest.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
Finished Running Test: testReadYourWritesBySize in 13.348 seconds.
09:28:34,452  INFO |testReadYourWritesByGet| - [NodeExtension] Thread-4800 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Destroying node NodeExtension.
09:28:34,453  INFO |testReadYourWritesByGet| - [Node] Thread-4800 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Hazelcast Shutdown is completed in 96 ms.
09:28:34,453  INFO |testReadYourWritesByGet| - [LifecycleService] Thread-4800 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after testReadYourWritesByGet(com.hazelcast.replicatedmap.ReplicatedMapReadYourWritesTest): BuildInfo{version='5.0-SNAPSHOT', build='20210428', buildNumber=20210428, revision=a214661, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='5.0-SNAPSHOT', build='20210428', revision='a214661'}}
Hiccups measured while running test 'testReadYourWritesByGet(com.hazelcast.replicatedmap.ReplicatedMapReadYourWritesTest):'
09:28:20, accumulated pauses: 836 ms, max pause: 64 ms, pauses over 1000 ms: 0
09:28:25, accumulated pauses: 7620 ms, max pause: 7327 ms, pauses over 1000 ms: 1
09:28:30, accumulated pauses: 459 ms, max pause: 220 ms, pauses over 1000 ms: 0
```

I tried to reproduce the failure, but no success. However, here is my opinion of the test failure. In `ReplicatedMapReadYourWritesTest#testReadYourWritesByGet` we generate a key that belongs to local instance, call write operation for that key, and then read it back. By the [definition / implementation](https://docs.hazelcast.com/imdg/4.2/data-structures/replicated-map.html) of `ReplicatedMap` writes are always forwarded to the partition owner and reads are done locally. 

For the `ReplicatedMapReadYourWritesTest#testReadYourWritesByGet` as long as there is no cluster state change, everything works well, i.e., the test failure above is due to cluster state change when tests are run in parallel.

Here is sample test class, when executed one method can affect the other one, i.e., tests are not run in complete isolation


    @RunWith(HazelcastParallelClassRunner.class)
    @Category({QuickTest.class, ParallelJVMTest.class})
    public class ReplicatedMapHypoTest extends ReplicatedMapAbstractTest {

      private static final TestHazelcastInstanceFactory nodeFactory = new TestHazelcastInstanceFactory();

      private static final BlockingQueue<Integer> DOOR = new ArrayBlockingQueue<>(1);
  
  
      @AfterClass
      public static void shutdown() {
          nodeFactory.shutdownAll();
      }

  
      @Test
      public void testSimulateCreateDefaultReplicatedMap() throws InterruptedException {
          DOOR.take();
  
          final int n = 10;
  
          List<HazelcastInstance> instances = new ArrayList<>(n);
          for (int i = 0; i < n; ++i) {
              instances.add(nodeFactory.newHazelcastInstance());
          }
  
          DOOR.take();
      }
  
      @Test
      public void testPartitionOwnership() throws InterruptedException {
          final int n = 10;
  
          List<HazelcastInstance> instances = new ArrayList<>(n);
          for (int i = 0; i < n; ++i) {
              instances.add(nodeFactory.newHazelcastInstance());
          }
  
          HazelcastInstance firstInstance = instances.get(0);
          String keyOwnedByInstance1 = generateKeyOwnedBy(firstInstance);
  
          DOOR.put(1);
          TimeUnit.SECONDS.sleep(50);
  
          assertEquals(firstInstance.getPartitionService().getPartition(keyOwnedByInstance1).getOwner(),
                  firstInstance.getCluster().getLocalMember());
  
  
          DOOR.put(1);
      }
    }
 

And we get the following error:

    [ERROR] Failures: 
    [ERROR]     ReplicatedMapHypoTest.testPartitionOwnership:73 expected:<Member [127.0.0.1]:5719 - e5de9f90-6dc6-4c5e-97b5-10f66e74ca73> but was:<Member [127.0.0.1]:5701 - 460f53df-4690-4db4-afb6-8caf52c42d90 this>


And here is a plausible sequence of events that cause test failure in the issue: 

1. We have `HazelcastInstance instance1, instance2` and `ReplicatedMap map1, map2` where `map1 = (same distributed objects) = map2`
2. Generate `KEY` for which `instance1` is the owner.
3. `HazelcastInstance instance3` from another parallel running test having `map3` joins the cluster, where `map1 == map2 == map3`
4. Now, `instance3` is the owner of the partition where `KEY` belongs
5. `instance1` gets a write request for `KEY` and redirects it to `instance3`
6. `instance3` returns success, while not having replicated yet `KEY` to all other members (since this is a normal behavior for `ReplicatedMap`)
7. `instance1` returns success
8. `<read> KEY` request is sent to `instance1`
9. `instance1` has not yet received the update from `instance3`, and returns `NULL`
10. Read your writes is violated

--- a/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapReadYourWritesTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/replicatedmap/ReplicatedMapReadYourWritesTest.java
@@ -67,8 +67,8 @@ public class ReplicatedMapReadYourWritesTest extends ReplicatedMapAbstractTest {
         ReplicatedMap<String, Integer> map1 = instance1.getReplicatedMap("default");
         ReplicatedMap<String, Integer> map2 = instance2.getReplicatedMap("default");
         for (int i = 0; i < 1000; i++) {
-            assertReadYourWriteByGet(instance2, map1, i);
-            assertReadYourWriteByGet(instance1, map2, i);
+            assertEventuallyReadYourWriteByGet(instance2, map1, i);
+            assertEventuallyReadYourWriteByGet(instance1, map2, i);
         }
     }
 
@@ -80,8 +80,8 @@ public class ReplicatedMapReadYourWritesTest extends ReplicatedMapAbstractTest {
         ReplicatedMap<String, Integer> map1 = instance1.getReplicatedMap("default");
         ReplicatedMap<String, Integer> map2 = instance2.getReplicatedMap("default");
         for (int i = 0; i < 1000; i++) {
-            assertReadYourWriteByContainsKey(instance2, map1, i);
-            assertReadYourWriteByContainsKey(instance1, map2, i);
+            assertEventuallyReadYourWriteByContainsKey(instance2, map1, i);
+            assertEventuallyReadYourWriteByContainsKey(instance1, map2, i);
         }
     }
 
@@ -93,25 +93,39 @@ public class ReplicatedMapReadYourWritesTest extends ReplicatedMapAbstractTest {
         ReplicatedMap<String, Integer> map1 = instance1.getReplicatedMap("default");
         ReplicatedMap<String, Integer> map2 = instance2.getReplicatedMap("default");
         for (int i = 0; i < 1000; i++) {
-            assertReadYourWriteByContainsValue(instance2, map1, i);
-            assertReadYourWriteByContainsValue(instance1, map2, i);
+            assertEventuallyReadYourWriteByContainsValue(instance2, map1, i);
+            assertEventuallyReadYourWriteByContainsValue(instance1, map2, i);
         }
     }
 
-
-    private void assertReadYourWriteByGet(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
+    private void assertEventuallyReadYourWriteByGet(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
         String key = generateKeyAndPutValue(instance, map, value);
-        assertEquals(value, (int) map.get(key));
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertEquals(value, (int) map.get(key));
+            }
+        });
     }
 
-    private void assertReadYourWriteByContainsKey(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
+    private void assertEventuallyReadYourWriteByContainsKey(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
         String key = generateKeyAndPutValue(instance, map, value);
-        assertTrue(map.containsKey(key));
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertTrue(map.containsKey(key));
+            }
+        });
     }
 
-    private void assertReadYourWriteByContainsValue(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
+    private void assertEventuallyReadYourWriteByContainsValue(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {
         generateKeyAndPutValue(instance, map, value);
-        assertTrue(map.containsValue(value));
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() throws Exception {
+                assertTrue(map.containsValue(value));
+            }
+        });
     }
 
     private String generateKeyAndPutValue(HazelcastInstance instance, ReplicatedMap<String, Integer> map, int value) {

